### PR TITLE
user-following-part1-user

### PR DESCRIPTION
**Pull request by phamhuycuong**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,22 +1,61 @@
 class UsersController < ApplicationController
-  
+  before_action :set_user, except: [:show, :new, :index, :create]
```
Comment (khanhkhanh) [inline app/controllers/users_controller.rb]: Use only instead of except

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,22 +1,61 @@
...
+    if @user.update user_params
+      flash[:success] = I18n.t "user.update_success"
+      redirect_to root_path
+    else
+      render "edit"
+    end
+  end
+
+  def destroy
+    @user.destroy
```
Comment (khanhkhanh) [inline app/controllers/users_controller.rb]: if ... else ... end

File: app/models/user.rb [hunk 3]
```diff
@@ -1,14 +1,38 @@
...
-  validates :email, length: {minimum:Settings.user_email_min_long, too_short: I18n.t(:user_error_too_short)}, uniqueness: true  
-    
+  validates :name, length: {minimum: Settings.user_name_min_long, too_short: I18n.t(:user_error_too_short)}, uniqueness: true
+
+  validates :email, length: {minimum:Settings.user_email_min_long, too_short: I18n.t(:user_error_too_short)}, uniqueness: true
+
   validates :password, length: {minimum: Settings.user_password_min_long, too_short: I18n.t(:user_error_too_short)}
+
+  def follow other_user
+    active_relationships.create(followed_id: other_user.id)
```
Comment (khanhkhanh) [inline app/models/user.rb]: Remove ()

File: app/views/users/_lesson.html.erb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    <% user.lessons.each do |lesson| %>
+      <tr>
+        <td>
+          <h4><%= t "user.learned_lesson_in_category", category_title: lesson.category.title.capitalize %></h4>
+          <em>
+            <%= t "user.time_passed", time: time_ago_in_words(lesson.created_at), passed: lesson.category.words.lesson_passed(lesson).count  %>
+          </em>
+        </td>
+        <td align="right">
+          <h4>&nbsp</h4>
```
Comment (khanhkhanh) [inline app/views/users/_lesson.html.erb]: &nbsp => &nbsp;

File: app/views/users/_lesson.html.erb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+          <em>
+            <%= link_to_if(current_user.admin?, t(:destroy), category_lesson_path(lesson, category_id: lesson.category.id), method: :delete, data: {confirm: t(:are_you_sure)}){} %>
```
Comment (khanhkhanh) [inline app/views/users/_lesson.html.erb]: Try removing ()
link_to_if(current_user.admin?, t(:destroy), category_lesson_path(lesson, category_id: lesson.category.id), method: :delete, data: {confirm: t(:are_you_sure)})
=> link_to_if current_user.admin?, t(:destroy), category_lesson_path(lesson, category_id: lesson.category.id), method: :delete, data: {confirm: t(:are_you_sure)}

File: app/views/users/_unfollow.html.erb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: user.id),
```
Comment (khanhkhanh) [inline app/views/users/_unfollow.html.erb]: Remove ()

File: app/views/users/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+<h1>Listing Users</h1>
```
Comment (khanhkhanh) [inline app/views/users/index.html.erb]: i18n

File: app/views/users/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  <div class="col-md-8" align="center">
+    <% if @users.any? %>
+      <table class="table table-hover w100">
+        <tr>
+          <th width="5%"><%= t "user.id" %></th>
+          <th><%= t "user.name" %></th>
+          <th><%= t "user.email" %></th>
+          <td colspan="4"></td>
+        </tr>
+        <% @users.each do |user| %>
```
Comment (khanhkhanh) [inline app/views/users/index.html.erb]: Use render

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -1,22 +1,59 @@
...
+  before_action :check_login, except: [:new, :create]
+
+  def index
+    if params[:status].present?
+      @user = User.friendly.find params[:id]
+      @follows = @user.send(params[:status]).paginate page: params[:page]
+      @title = t "user." + params[:status], count: @follows.count
+      render "show_follow"
+    else
+      @users = User.all
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Pagination

File: app/helpers/users_helper.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+module UsersHelper
+  def gravatar_for user, options = {size: Settings.user.gravantar_size90}
+    gravatar_id = Digest::MD5::hexdigest user.email.downcase
+    size = options[:size]
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}?s=#{size}"
```
Comment (phamkykhoi) [inline app/helpers/users_helper.rb]: Should use upload file from local

File: app/views/users/_unfollow.html.erb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<%= form_for current_user.active_relationships.find_by(followed_id: user.id),
+             html: { method: :delete } do |f| %>
```
Comment (phamkykhoi) [inline app/views/users/_unfollow.html.erb]: Remove space {  }

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -1,22 +1,59 @@
...
+    if @user.update user_params
+      flash[:success] = I18n.t "user.update_success"
+      redirect_to root_path
+    else
+      render "edit"
+    end
+  end
+
+  def destroy
+    @user.destroy ? flash[:success] = I18n.t("user.destroy_success") : flash[:danger] = I18n.t("user.destroy_error")
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: `if..else`

File: app/controllers/relationships_controller.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
+class RelationshipsController < ApplicationController
+
+  def index
+    if params[:status].present?
```
Comment (dieunb) [inline app/controllers/relationships_controller.rb]: missing condition, if params[:status] is not `following` and `followers`

File: app/views/users/_lesson.html.erb [hunk 14]
```diff
@@ -0,0 +1,20 @@
+<% if user.lessons.any? %>
+  <table class="table table-hover w100">
+    <% user.lessons.each do |lesson| %>
```
Comment (dieunb) [inline app/views/users/_lesson.html.erb]: refactor loop with partial file

File: app/controllers/relationships_controller.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
+class RelationshipsController < ApplicationController
+
+  def index
+    if params[:status].present?
+      @user = User.friendly.find params[:id]
+      @follows = @user.send(params[:status]).paginate page: params[:page], per_page: Settings.user.follow_paginate
+      @title = t "user." + params[:status], count: @follows.count
+      render "users/show_follow"
```
Comment (khanhkhanh) [inline app/controllers/relationships_controller.rb]: Thay vì anh để show_follow trong views/users, anh chuyển file này qua thư mục views/relationships và đổi tên thành index.html.erb

File: config/routes.rb [hunk 16]
```diff
@@ -8,13 +8,21 @@
 
   get "signup"  => "users#new"
 
-  resources :users
   resources :words
+  resources :users do
+    resources :following, only: [:show]
```
Comment (dieunb) [inline config/routes.rb]: No need.

File: config/routes.rb [hunk 17]
```diff
@@ -8,13 +8,21 @@
...
+    resources :followers, only: [:show]
```
Comment (dieunb) [inline config/routes.rb]: No need.

File: config/routes.rb [hunk 18]
```diff
@@ -8,13 +8,21 @@
...
+  end
 
   resources :categories do
     resources :lessons
   end
 
+  resources :relationships, only: [:index, :create, :destroy]
+
+  get "/users/:id/following" => "relationships#index", as: "following"
+  get "/users/:id/followers" => "relationships#index", as: "followers"
```
Comment (dieunb) [inline config/routes.rb]: Merge 2 routes to one routes use sending params

File: app/views/users/_unfollow.html.erb [hunk 19]
```diff
@@ -0,0 +1,4 @@
+<%= form_for current_user.active_relationships.find_by(followed_id: user.id),
+             html: {method: :delete} do |f| %>
```
Comment (khanhkhanh) [inline app/views/users/_unfollow.html.erb]: indent

File: app/views/users/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,10 @@
+<div class="row">
+  <div class="col-md-3">
+    <%= render "users/summary", user: @user %>
+  </div>
+  <div class="col-md-9">
+    <%= @user == current_user ?  render("shared/word_lesson_button") : render("follow_form", user: @user) %>
```
Comment (khanhkhanh) [inline app/views/users/show.html.erb]: 2 spaces after ?

File: config/routes.rb [hunk 21]
```diff
@@ -8,13 +8,18 @@
...
+  resources :users
 
   resources :categories do
     resources :lessons
   end
 
+  resources :relationships, only: [:index, :create, :destroy]
+
+  get "/users/:id/following" => "relationships#index"
+  get "/users/:id/followers" => "relationships#index"
```
Comment (dieunb) [inline config/routes.rb]: merge 2 routes to one routes

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -1,22 +1,57 @@
...
+    end
+  end
+
+  def destroy
+    if @user.destroy
+      flash[:success] = I18n.t "user.destroy_success"
+      redirect_to users_path
+    else
+      flash[:danger] = I18n.t "user.destroy_error"
+      redirect_to users_path
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: duplicated `redirect_to users_path`

File: config/routes.rb [hunk 23]
```diff
@@ -8,13 +8,17 @@
...
   resources :words
+  resources :users
 
   resources :categories do
     resources :lessons
   end
 
+  resources :relationships, only: [:index, :create, :destroy]
+
+  get "/users/:id/follows"  => "relationships#index", as: "follows"
```
Comment (dieunb) [inline config/routes.rb]: +  get "/users/:id/:type"  => "relationships#index", as: "friendship"

**Pull request merged**